### PR TITLE
Turn on CodeQL and fix BinSkim regressions

### PR DESCRIPTION
**Pull request by BillyONeal**

### CODE REVIEW

**Review by vicroms — APPROVED** [review]

File: cmake/Findfmt.cmake [hunk 1]
```diff
@@ -12,6 +12,23 @@ if(NOT VCPKG_FMT_URL)
...
+if(MSVC AND VCPKG_DEVELOPMENT_WARNINGS AND NOT (CMAKE_CXX_COMPILER_ID MATCHES "AppleClang") AND NOT (CMAKE_CXX_COMPILER_ID MATCHES "[Cc]lang"))
+    set(SKIP_WARNINGS ON)
+    # fmt\core.h(418): warning C6239: (<non-zero constant> && <expression>) always evaluates to the result of <expression>:  Did you intend to use the bitwise-and (&) operator? If not, consider removing the redundant '<non-zero constant>' and the && operator.
+    string(APPEND CMAKE_CXX_FLAGS " /wd6239")
+    # This one is guarded by an assert
+    # fmt\format-inl.h(327): warning C6385: Reading invalid data from 'pow10_significands'.: Lines: 298, 300, 327
+    string(APPEND CMAKE_CXX_FLAGS " /wd6385")
+    # fmt\os.h(377): warning C6326: Potential comparison of a constant with another constant.
+    string(APPEND CMAKE_CXX_FLAGS " /wd6326")
+endif()
```
Comment (ras0219-msft) [inline cmake/Findfmt.cmake:30]: CMake already has a system of "directory-scoped flags": `add_compile_options()`
```suggestion
if(MSVC AND VCPKG_DEVELOPMENT_WARNINGS AND NOT (CMAKE_CXX_COMPILER_ID MATCHES "AppleClang") AND NOT (CMAKE_CXX_COMPILER_ID MATCHES "[Cc]lang"))
    # fmt\core.h(418): warning C6239: (<non-zero constant> && <expression>) always evaluates to the result of <expression>:  Did you intend to use the bitwise-and (&) operator? If not, consider removing the redundant '<non-zero constant>' and the && operator.
    add_compile_options(/wd6239)
    # This one is guarded by an assert
    # fmt\format-inl.h(327): warning C6385: Reading invalid data from 'pow10_significands'.: Lines: 298, 300, 327
    add_compile_options(/wd6385)
    # fmt\os.h(377): warning C6326: Potential comparison of a constant with another constant.
    add_compile_options(/wd6326)
endif()
```

**Comment by Neumann-A** [comment]
How about using CMake presets instead and be a good example how CMake should be done ?

File: CMakeLists.txt [hunk 2]
```diff
@@ -116,10 +116,54 @@ set(CMAKE_CXX_EXTENSIONS OFF)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_STANDARD 17)
 if(MSVC)
-    string(APPEND CMAKE_CXX_FLAGS " /EHsc")
-    if(CMAKE_BUILD_TYPE STREQUAL "Release")
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zi /guard:cf")
-        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /DEBUG /debugtype:cv,fixup /guard:cf")
+    # either MSVC, or clang-cl
```
Comment (ras0219-msft) [inline CMakeLists.txt:119]: In general, we should either:
1. `add_compile_options()` / `add_link_options()`
2. Move these to a Toolchain file

Modifying CMAKE_C(XX)_FLAGS in the CMakeLists.txt is always a smell. That said, this works and follows what came before, so not requiring changes at this time.

**Review by ras0219-msft — APPROVED** [review]

File: cmake/Findfmt.cmake (hunk 1 above)
Comment (BillyONeal) [reply cmake/Findfmt.cmake:30]: This isn't "in a directory" though, it needs to edit the cache variables that are passed down through `FetchContent_Declare` -> `ExternalProject_Add`.

**Comment by BillyONeal** [comment]
> How about using CMake presets instead and be a good example how CMake should be done ?
> [...]
> In general, we should either:
> 
> 1. `add_compile_options()` / `add_link_options()`
> 2. Move these to a Toolchain file
> 
> Modifying CMAKE_C(XX)_FLAGS in the CMakeLists.txt is always a smell. That said, this works and follows what came before, so not requiring changes at this time.

This is a controversial opinion among the vcpkg maintainers, but I still believe that CMake expecting application authors to separate warnings manipulation into separate files is unreasonable. And I cite as evidence that I've reviewed a lot of things that are shipped into vcpkg ports and never ever seen a project do it. Separate files is a statement that separate people are interested in the values, and that isn't true for diagnostics.

Contrast with `/machine:` switches in similar that make sense to be part of the "toolchain".

File: cmake/Findfmt.cmake (hunk 1 above)
Comment (BillyONeal) [reply cmake/Findfmt.cmake:30]: Interestingly, this created the opposite problem. It did indeed reach down into the `ExternalProject_Add`, but it also applied the effect to vcpkg itself which is undesired:

image

**Comment by Neumann-A** [comment]
> And I cite as evidence that I've reviewed a lot of things that are shipped into vcpkg ports and never ever seen a project do it.

Most ports predate presets and most projects don't care about proper CMake ;)
VTK/Qt in general give an option to opt out of the flag setting business.

**Comment by BillyONeal** [comment]
>VTK/Qt in general give an option to opt out of the flag setting business.

And so do we. `VCPKG_DEVELOPMENT_WARNINGS` defaults to off.

**Comment by Neumann-A** [comment]
> And so do we. VCPKG_DEVELOPMENT_WARNINGS defaults to off.

That is just the warning flag setting business and leaves at least:
https://github.com/microsoft/vcpkg-tool/blob/5fdee72bc1fceca198fb1ab7589837206a8b81ba/CMakeLists.txt#L120-L130

**Pull request merged**